milkv/pioneer/linuxboot-initrd.nix

```diff
@@ -0,0 +1,54 @@
+{ buildGoModule
+, fetchFromGitHub
+, fetchpatch
+, linux-firmware
+, ...
+}:
+
+# Based on
+# https://github.com/sophgo/bootloader-riscv/blob/e0839852d571df106db622611f4786ae17e8df0f/scripts/envsetup.sh#L809-L819
+
+buildGoModule rec {
+  pname = "u-root";
+  version = "0.14.0";
+  src = fetchFromGitHub {
+    owner = "u-root";
+    repo = "u-root";
+    rev = "v${version}";
+    hash = "sha256-8zA3pHf45MdUcq/MA/mf0KCTxB1viHieU/oigYwIPgo=";
+  };
+  vendorHash = null;
+  patches = [
+    (
+      fetchpatch {
+        url = "https://github.com/sophgo/bootloader-riscv/commit/322c3305763872a9b88a1c85d79bca63b8fbe7a6.patch";
+        hash = "sha256-l5r3DbcMqRYD5FhRBqtEIEscZAdDvjmQJE4BIAtWYWE=";
+        stripLen = 1;
+      }
+    )
+  ];
+
+  # We only build the u-root binary in the build phase and the initrd in the
+  # postBuild hook.
+  subPackages = [ "." ];
+  postBuild = ''
+    GOROOT="$(go env GOROOT)" $GOPATH/bin/u-root \
```

Comment:
This doesn't properly cross-compile:

```
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/rzqv74kh941ifkqgrs3x44w5cm0bb2g8-u-root-riscv64-unknown-linux-gnu-0.14.0.drv^*'
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/zsnkg41sqpw8zaw4jhcyasnpy2nzx1v9-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/9x0d3i08pgdq3lsj5029gl6l0yflzlcv-322c3305763872a9b88a1c85d79bca63b8fbe7a6.patch
patching file pkg/boot/syslinux/syslinux.go
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
Building subPackage ./.
/nix/store/wpxb5d446br115xxcv6gi7f8nkh2i5x0-stdenv-linux/setup: line 114: /build/go/bin/u-root: cannot execute binary file: Exec format error
/nix/store/wpxb5d446br115xxcv6gi7f8nkh2i5x0-stdenv-linux/setup: line 131: pop_var_context: head of shell_variables not a function context
```

Reply:
I had binfmt emulation enabled and didn't notice that. I get the same error without the emulation.

Reply:
I have a fix but not yet tested on hardware.

Reply:
From a341d5c5e29c106ed45e5a1308ce591daed96c54 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=B6rg=20Thalheim?= <joerg@thalheim.io>
Date: Sun, 7 Apr 2024 17:42:03 +0000
Subject: [PATCH] fix u-root build
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Jörg Thalheim <joerg@thalheim.io>
---
 milkv/pioneer/linuxboot-initrd.nix | 77 +++++++++++++++++++-----------
 1 file changed, 48 insertions(+), 29 deletions(-)

diff --git a/milkv/pioneer/linuxboot-initrd.nix b/milkv/pioneer/linuxboot-initrd.nix
index ee0ef7b..507fc0e 100644
--- a/milkv/pioneer/linuxboot-initrd.nix
+++ b/milkv/pioneer/linuxboot-initrd.nix
@@ -1,54 +1,73 @@
-{ buildGoModule
+{ buildPackages
 , fetchFromGitHub
 , fetchpatch
 , linux-firmware
+, buildGoModule
 , ...
 }:
 
 # Based on
 # https://github.com/sophgo/bootloader-riscv/blob/e0839852d571df106db622611f4786ae17e8df0f/scripts/envsetup.sh#L809-L819
-
-buildGoModule rec {
-  pname = "u-root";
-  version = "0.14.0";
-  src = fetchFromGitHub {
-    owner = "u-root";
-    repo = "u-root";
-    rev = "v${version}";
-    hash = "sha256-8zA3pHf45MdUcq/MA/mf0KCTxB1viHieU/oigYwIPgo=";
-  };
-  vendorHash = null;
-  patches = [
-    (
-      fetchpatch {
+let
+  u-root = buildPackages.buildGoModule rec {
+    pname = "u-root";
+    version = "0.14.0";
+    src = fetchFromGitHub {
+      owner = "u-root";
+      repo = "u-root";
+      rev = "v${version}";
+      hash = "sha256-8zA3pHf45MdUcq/MA/mf0KCTxB1viHieU/oigYwIPgo=";
+    };
+    vendorHash = null;
+    patches = [
+      (fetchpatch {
         url = "https://github.com/sophgo/bootloader-riscv/commit/322c3305763872a9b88a1c85d79bca63b8fbe7a6.patch";
         hash = "sha256-l5r3DbcMqRYD5FhRBqtEIEscZAdDvjmQJE4BIAtWYWE=";
         stripLen = 1;
-      }
-    )
-  ];
+      })
+    ];
+
+    postInstall = ''
+      cp -r . $out/src
+    '';
+
+    # We only build the u-root binary in the build phase and the initrd in the
+    # postBuild hook.
+    subPackages = [ "." ];
 
-  # We only build the u-root binary in the build phase and the initrd in the
-  # postBuild hook.
-  subPackages = [ "." ];
-  postBuild = ''
-    GOROOT="$(go env GOROOT)" $GOPATH/bin/u-root \
+    # Tests time out after 10min for native riscv64 builds on the pioneer.
+    doCheck = false;
+  };
+in
+buildGoModule {
+  name = "linuxboot-initrd";
+  src = null;
+  vendorHash = null;
+  dontUnpack = true;
+  nativeBuildInputs = [ u-root ];
+
+  buildPhase = ''
+    runHook preBuild
+    pushd ${u-root}/src
+    mkdir -p $out
+    GOROOT="$(go env GOROOT)" u-root \
       -build bb \
       -uinitcmd=boot \
       -files "${linux-firmware}/lib/firmware/amdgpu/:lib/firmware/amdgpu/" \
       -files "${linux-firmware}/lib/firmware/radeon/:lib/firmware/radeon/" \
-      -o initramfs.cpio \
+      -o $out/initramfs.cpio \
       core boot
+    popd
 
     # The vendor does not compress the initrd. We do since we include more
     # firmware files. CRC32 is required by the kernel's decompressor.
-    xz --check=crc32 initramfs.cpio
+    xz --check=crc32 $out/initramfs.cpio
+    runHook postBuild
   '';
 
   installPhase = ''
-    install -D initramfs.cpio.xz $out/initrd.img
+    runHook preInstall
+    mv $out/initramfs.cpio.xz $out/initrd.img
+    runHook postInstall
   '';
-
-  # Tests time out after 10min for native riscv64 builds on the pioneer.
-  doCheck = false;
 }
-- 
2.44.0

Reply:
https://github.com/Mic92/nixos-hardware/commit/9c53953efab156e983eca5338adafec29b0a938e

Reply:
I also had to use staging because in master there is a fix for cross compilation missing.

Reply:
Pushed.